src/NHibernate/Util/SerializationHelper.Surrogates.cs

```diff
@@ -0,0 +1,312 @@
+			}
+
+			public object SetObjectData(object obj, SerializationInfo info, StreamingContext context, ISurrogateSelector selector) =>
+				throw new NotSupportedException();
+		}
+
+		[Serializable]
+		private sealed class DelegateReference : IObjectReference, ISerializable
+		{
+			private readonly System.Type[] _types;
+			private readonly MethodInfo[] _methods;
+			private readonly object[] _targets;
+
+			public DelegateReference(Delegate @delegate)
+			{
+				if (@delegate == null)
+					throw new ArgumentNullException(nameof(@delegate));
+
+				var invocations = @delegate.GetInvocationList();
+				_types = new System.Type[invocations.Length];
+				_methods = new MethodInfo[invocations.Length];
+				_targets = new object[invocations.Length];
+				for (var i = 0; i < invocations.Length; i++)
+				{
+					var invocation = invocations[i];
+					_types[i] = invocation.GetType();
+					_methods[i] = invocation.Method;
+					_targets[i] = invocation.Target;
+				}
+			}
+
+			private DelegateReference(SerializationInfo info, StreamingContext context)
+			{
+				_types = GetValueArray<System.Type>(info, "Types");
+				_methods = info.GetValueArray<MethodInfo>("Methods");
+				_targets = info.GetValueArray<object>("Targets");
+			}
+
+			public void GetObjectData(SerializationInfo info, StreamingContext context)
+			{
+				info.AddValueArray("Types", _types);
+				info.AddValueArray("Methods", _methods);
+				info.AddValueArray("Targets", _targets);
+			}
+
+			public object GetRealObject(StreamingContext context)
+			{
+				var delegates = new Delegate[_types.Length];
+				for (var i = 0; i < delegates.Length; i++)
+					delegates[i] = Delegate.CreateDelegate(_types[i], _targets[i], _methods[i]);
+
+				return delegates.Length == 0
+					? delegates[0]
+					: Delegate.Combine(delegates);
+			}
+		}
+
+		public sealed class SurrogateSelector : ISurrogateSelector
+		{
+			private static readonly bool SystemTypeIsSerializable = typeof(System.Type).IsSerializable;
+			private static readonly bool MemberInfoIsSerializable = typeof(MemberInfo).IsSerializable;
+			private static readonly bool DelegateIsSerializable = GetDelegateIsSerializable();
+
+			private static bool GetDelegateIsSerializable()
+			{
+				//In .NET Standard & .NET Core Delegate implements ISerializable,
+				//but throws SerializationException
+				if (!typeof(Delegate).IsSerializable)
+					return false;
+				try
+				{
+					System.Action a = () => { };
+					a.GetObjectData(
+						new SerializationInfo(typeof(Delegate), new FormatterConverter()),
+						new StreamingContext(StreamingContextStates.All));
+					return true;
+				}
+				catch (PlatformNotSupportedException)
+				{
+					return false;
+				}
+				catch (SerializationException)
+				{
+					return false;
+				}
+			}
+
+			private ISurrogateSelector _next;
+
+			void ISurrogateSelector.ChainSelector(ISurrogateSelector selector)
+				=> _next = selector;
```

Comment:
The interface contract is not explicit, but looking at [`System.Runtime.Serialization.SurrogateSelector` code](https://github.com/Microsoft/referencesource/blob/4.6.2/mscorlib/system/runtime/serialization/surrogateselector.cs#L81), it seems it is supposed to handle multiple chaining without losing the previous one as does the code here. (It also does extensive checking for preventing cycles.)

The logic for chaining is to set `selector` as the next one and if there was already another next, then chain it to the last next of `selector` (recursive call to GetNextSelector until finding null, then chaining to the last not null).

Rather than re-coding this it would be better if `SurrogateSelector` implementation could be reused. Otherwise maybe harden the checks here for stating our own contract, like throwing if `_next` is not null and tell to chain on `GetNextSelector` instead.

Reply:
It is possible to just inherit SurrogateSelector.